src/librustc_mir/transform/match_branches.rs

```diff
@@ -48,7 +48,7 @@ impl<'tcx> MirPass<'tcx> for MatchBranchSimplification {
                     ref targets,
                     ref values,
                     ..
-                } if targets.len() == 2 && values.len() == 1 => {
+                } if targets.len() == 2 && values.len() == 1 && targets[0] != targets[1] => {
```

Comment:
Just curious, is targets[0] == targets[1] ever hit? I imagine that would be a whole different opt pass

Reply:
I don't know. 

The CfgSimplifier optimizes switch with the same targets to a goto, and then collapses goto chains and merges blocks.

Reply:
we can make this condition an assert after the beta branch, then we have 12 weeks to find out if it's a problem :D